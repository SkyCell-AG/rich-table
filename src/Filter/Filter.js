import React from 'react'
import PropTypes from 'prop-types'
import get from 'lodash/get'
import CircularProgress from '@material-ui/core/CircularProgress'
import List from '@material-ui/core/List'
import TextField from '@material-ui/core/TextField'

import FilterPopover from '../FilterPopover'
import * as statuses from 'utils/requestStatuses'
import getDataFromInput from 'utils/getDataFromInput'

import FilterItem from './FilterItem'

import useStyles from './Filter.styles'

const propTypes = {
    className: PropTypes.string,
    button: PropTypes.element,
    id: PropTypes.string.isRequired,
    setSearchPhrase: PropTypes.func.isRequired,
    searchPhrase: PropTypes.string.isRequired,
    filter: PropTypes.arrayOf(PropTypes.string),
    set: PropTypes.func.isRequired,
    load: PropTypes.func.isRequired,
    predefinedFilter: PropTypes.object, // eslint-disable-line
    filters: PropTypes.array, // eslint-disable-line
    status: PropTypes.oneOf([
        statuses.SUCCESS,
        statuses.FAILURE,
        statuses.PENDING,
        statuses.PRISTIN,
    ]).isRequired,
}

const defaultProps = {
    predefinedFilter: {},
    filter: [],
    filters: [],
    className: '',
    button: null,
}

const rearangeItems = (filterItems) => {
    const checkedItems = filterItems.filter(({
        checked,
    }) => {
        return checked
    }).sort((a, b) => {
        return a.label > b.label ? 1 : -1
    })

    const notCheckedItems = filterItems.filter(({
        checked,
    }) => {
        return !checked
    })

    return [
        ...checkedItems,
        ...notCheckedItems,
    ]
}

const Filter = (props) => {
    const {
        id,
        searchPhrase,
        predefinedFilter,
        setSearchPhrase,
        status,
        load,
        filters,
        set,
        filter,
        button,
        className,
    } = props

    const classes = useStyles()

    const filterItems = rearangeItems(
        filters
            .filter(({
                label = '',
                value = '',
            }) => {
                return (
                    value.toLowerCase().includes(searchPhrase.toLowerCase())
                    || label.toLowerCase().includes(searchPhrase.toLowerCase())
                )
            })
            .map((itemProps) => {
                const {
                    value: itemValue,
                } = itemProps
                const selected = filter.includes(itemValue)
                const predefinedSelected = get(predefinedFilter, `[${id}]`, []).includes(itemValue)

                return {
                    ...itemProps,
                    checked: selected,
                    disabled: predefinedSelected,
                    onClick: (e) => {
                        e.stopPropagation()
                        e.preventDefault()
                        if (predefinedSelected) {
                            return
                        }

                        set(selected ? (
                            filter.filter((item) => {
                                return item !== itemValue
                            })
                        ) : (
                            [
                                ...filter,
                                itemValue,
                            ]
                        ))
                    },
                }
            }),
    )

    return (
        <FilterPopover
            className={className}
            onOpen={load}
            active={filter.length > 0}
            button={button}
        >
            <div className={classes.filterListContainer}>
                {status === statuses.FAILURE && (
                    <div>an error happened.</div>
                )}
                {(
                    status === statuses.PENDING || status === statuses.PRISTIN
                ) && (
                    <div className={classes.loader}>
                        <CircularProgress
                            size={40}
                        />
                    </div>
                )}
                {status === statuses.SUCCESS && (
                    <List
                        className={classes.filtersList}
                    >
                        <TextField
                            onChange={getDataFromInput(setSearchPhrase)}
                            className={classes.searchField}
                            label="Filter search"
                            value={searchPhrase}
                            autoFocus
                        />
                        {filterItems.map(FilterItem)}
                    </List>
                )}
            </div>
        </FilterPopover>
    )
}

Filter.propTypes = propTypes
Filter.defaultProps = defaultProps

export default Filter
