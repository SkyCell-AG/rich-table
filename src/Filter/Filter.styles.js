import {
    makeStyles,
} from '@material-ui/core'

const useStyles = makeStyles((theme) => {
    return {
        filterListContainer: {
            minWidth: 200,
            minHeight: 100,
            maxHeight: 400,
        },

        loader: {
            display: 'flex',
            justifyContent: 'center',
        },

        searchField: {
            display: 'flex',
            maxWidth: '100%',
            margin: `${theme.spacing(2)}px !important`,
        },
    }
})

export default useStyles
