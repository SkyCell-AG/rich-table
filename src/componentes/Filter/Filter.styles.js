import {
    makeStyles,
} from '@mui/styles'
import {
    useTheme,
} from '@mui/material/styles'

const useStyles = makeStyles(() => {
    const theme = useTheme()

    return {
        filterListContainer: {
            minWidth: 200,
            minHeight: 100,
            maxHeight: 400,
        },

        loader: {
            display: 'flex',
            justifyContent: 'center',
        },

        searchField: {
            display: 'flex',
            maxWidth: '100%',
            margin: `${theme.spacing(2)} !important`,
        },
    }
})

export default useStyles
