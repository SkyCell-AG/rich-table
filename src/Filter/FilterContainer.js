import React, {
    useReducer,
    useCallback,
} from 'react'
import PropTypes from 'prop-types'

import createReducer from 'utils/createReducer'
import * as statuses from 'utils/requestStatuses'
import generateAsyncActions from 'utils/generateAsyncActions'

import Filter from './Filter'

const INPUT_SEARCH_PHRASE = 'INPUT_SEARCH_PHRASE'
const LOAD_FILTERS = generateAsyncActions('LOAD_FILTERS')

const reducer = createReducer({
    [INPUT_SEARCH_PHRASE]: (state, {
        meta: {
            searchPhrase,
        },
    }) => {
        return {
            ...state,
            searchPhrase,
        }
    },
    [LOAD_FILTERS.pending]: (state) => {
        return {
            ...state,
            status: statuses.PENDING,
        }
    },
    [LOAD_FILTERS.failure]: (state, {
        err,
    }) => {
        return {
            ...state,
            status: statuses.FAILURE,
            err,
        }
    },
    [LOAD_FILTERS.success]: (state, {
        payload,
    }) => {
        return {
            ...state,
            status: statuses.SUCCESS,
            filters: payload,
        }
    },
})

const initState = {
    status: statuses.PRISTIN,
    err: null,
    filters: [],
    searchPhrase: '',
}

const propTypes = {
    loadFilters: PropTypes.func.isRequired,
}

const FilterContainer = (props) => {
    const {
        loadFilters,
    } = props

    const [
        state,
        dispatch,
    ] = useReducer(reducer, initState)

    const setSearchPhrase = useCallback((searchPhrase) => {
        dispatch({
            type: INPUT_SEARCH_PHRASE,
            meta: {
                searchPhrase,
            },
        })
    }, [])

    const load = useCallback(() => {
        dispatch({
            type: LOAD_FILTERS.pending,
        })
        loadFilters()
            .then((data) => {
                dispatch({
                    type: LOAD_FILTERS.success,
                    payload: data,
                })
            })
            .catch((err) => {
                dispatch({
                    type: LOAD_FILTERS.failure,
                    err,
                })
            })
    }, [loadFilters])

    return (
        <Filter
            {...props}
            searchPhrase={state.searchPhrase}
            setSearchPhrase={setSearchPhrase}
            status={state.status}
            filters={state.filters}
            load={load}
        />
    )
}

FilterContainer.propTypes = propTypes

export default FilterContainer
