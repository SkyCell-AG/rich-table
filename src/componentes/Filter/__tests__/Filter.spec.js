import React from 'react'
import {
    shallow,
    mount,
} from 'enzyme'

import Filter from '../Filter'

describe('app/shared-components/Filter should match snapshot', () => {
    it('should match snapshot', () => {
        const wrapper = shallow(
            <Filter
                className="makeStyles-item-2"
                id="containerStatus"
                setSearchPhrase={jest.fn()}
                searchPhrase=""
                filter={['9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832616UY']}
                set={jest.fn()}
                load={jest.fn()}
                predefinedFilter={{}}
                filters={[
                    {
                        count: 10,
                        label: '2500C',
                        value: '9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832616UY',
                    },
                    {
                        count: 4,
                        label: '1500C',
                        value: '9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832618B5',
                    },
                ]}
                status="SUCCESS"
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('renders without crashing', () => {
        mount(
            <Filter
                id="containerStatus"
                setSearchPhrase={jest.fn()}
                searchPhrase=""
                filter={['9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832616UY']}
                set={jest.fn()}
                load={jest.fn()}
                status="SUCCESS"
            />,
        )
    })
})
