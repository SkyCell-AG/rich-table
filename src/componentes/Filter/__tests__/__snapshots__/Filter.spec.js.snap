// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-components/Filter should match snapshot should match snapshot 1`] = `
<Component
  active={true}
  button={null}
  className="makeStyles-item-2"
  data-testid="filter-"
  onOpen={[MockFunction]}
>
  <div
    className="makeStyles-filterListContainer-1"
    data-testid="filter-container-"
  >
    <ForwardRef(List)>
      <ForwardRef(TextField)
        autoFocus={true}
        className="makeStyles-searchField-3"
        label="Filter search"
        onChange={[Function]}
        value=""
      />
      <FilterItem
        checked={true}
        count={10}
        disabled={false}
        key="9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832616UY-filterItem"
        label="2500C"
        onClick={[Function]}
        value="9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832616UY"
      />
      <FilterItem
        checked={false}
        count={4}
        disabled={false}
        key="9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832618B5-filterItem"
        label="1500C"
        onClick={[Function]}
        value="9x1EF20BD4D474D284ED257A1DAE235766F6EB3C2255BB0901832618B5"
      />
    </ForwardRef(List)>
  </div>
</Component>
`;
